gateway/src/tls.rs

```diff
@@ -0,0 +1,107 @@
+use std::collections::HashMap;
+use std::io::BufReader;
+use std::sync::Arc;
+
+use axum_server::accept::DefaultAcceptor;
+use axum_server::tls_rustls::{RustlsAcceptor, RustlsConfig};
+use futures::executor::block_on;
+use rustls::server::{ClientHello, ResolvesServerCert};
+use rustls::sign::{self, CertifiedKey};
+use rustls::{Certificate, PrivateKey, ServerConfig};
+use rustls_pemfile::Item;
+use shuttle_common::models::error::ErrorKind;
+use tokio::runtime::Handle;
+use tokio::sync::RwLock;
+
+use crate::service::GatewayService;
+use crate::{Error, Fqdn};
+
+pub fn parse_pem(certs: &[u8], key: &[u8]) -> Result<(Vec<Certificate>, PrivateKey), Error> {
+    let certs = rustls_pemfile::read_all(&mut BufReader::new(certs))
+        .map_err(|_| Error::from_kind(ErrorKind::Internal))
+        .and_then(|items| {
+            items
+                .into_iter()
+                .map(|item| match item {
+                    Item::X509Certificate(cert) => Ok(Certificate(cert)),
+                    _ => Err(Error::from_kind(ErrorKind::Internal)),
+                })
+                .collect()
+        })?;
+
+    let private_key = match rustls_pemfile::read_one(&mut BufReader::new(key)).unwrap() {
+        Some(Item::RSAKey(key)) | Some(Item::PKCS8Key(key)) | Some(Item::ECKey(key)) => {
+            Ok(PrivateKey(key))
+        }
+        _ => Err(Error::from_kind(ErrorKind::Internal)),
+    }?;
+
+    Ok((certs, private_key))
+}
+
+pub struct GatewayCertResolver {
+    keys: RwLock<HashMap<String, Arc<CertifiedKey>>>,
```

Comment:
Have you considered `dashmap` here? https://crates.io/crates/dashmap

I have to admit I'm not entirely sure what the performance trade-offs are in the Cert resolver - but it is a production ready drop-in replacement for `RwLock<HashMap<..>>` focusing on performance.

Reply:
Ah! I have seen this before. I feel I need to read a bit more into it before using it blindly though. Is the `RwLock` similar to `tokio`'s? I.e. write-preferring?

Reply:
Dashmap doesn't have an asynchronous API as far as I can tell, which may disqualify it immediately but I don't really have an intuition on this.

Here are some benchmarks of dashmap against other popular hashmap implementations (including `RwLock<HashMap>` but this `RwLock` in `std`).

I'm really not an expert on this crate - but from my understanding the philosophy of dashmap is to minimise contention which is achieved by sharding the hashmap internally:

```rust
pub struct DashMap<K, V, S = RandomState> {
    shift: usize,
    shards: Box<[RwLock<HashMap<K, V, S>>]>,
    hasher: S,
}
```

Reply:
Also found this: https://tokio.rs/tokio/tutorial/shared-state which I thought was interesting:

> Note, std::sync::Mutex and not tokio::sync::Mutex is used to guard the HashMap. A common error is to unconditionally use tokio::sync::Mutex from within async code. An async mutex is a mutex that is locked across calls to .await.

> A synchronous mutex will block the current thread when waiting to acquire the lock. This, in turn, will block other tasks from processing. However, switching to tokio::sync::Mutex usually does not help as the asynchronous mutex uses a synchronous mutex internally.

> As a rule of thumb, using a synchronous mutex from within asynchronous code is fine as long as contention remains low and the lock is not held across calls to .await. Additionally, consider using [parking_lot::Mutex](https://docs.rs/parking_lot/0.10.2/parking_lot/type.Mutex.html) as a faster alternative to std::sync::Mutex.

Reply:
Thank you! It's a lot of very useful info, I'll look into it 😄